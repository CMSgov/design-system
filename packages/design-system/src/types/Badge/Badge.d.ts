import * as React from 'react';

export type BadgeSize = 'big';

export type BadgeVariation = 'info' | 'success' | 'warn' | 'alert';

export interface BadgeProps {
  /**
   * Additional classes to be added to the root badge element.
   * Useful for adding utility classes.
   */
  className?: string;
  /**
   * Label text or HTML.
   */
  children: React.ReactNode;
  /**
   * Sets the font size of the Badge
   */
  size?: BadgeSize;
  /**
   * A string corresponding to the badge-component variation classes
   */
  variation?: BadgeVariation;
}

declare const Badge: React.FC<React.ComponentPropsWithRef<'span'> & BadgeProps>;

export default Badge;
