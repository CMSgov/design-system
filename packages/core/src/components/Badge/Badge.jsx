import React from 'react';

const Badge = (props) => {
  return <span className='ds-c-badge'>{props.children}</span>;
};

Badge.displayName = 'Badge';
Badge.propTypes = {
  /**
   * In most cases this will be the badge's label, but you could also use this
   * to nest more advanced JSX.
  */
  children: React.PropTypes.oneOfType([
    React.PropTypes.node,
    React.PropTypes.string
  ]).isRequired
};

export default Badge;
