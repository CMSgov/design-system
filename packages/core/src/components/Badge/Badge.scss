@import '@cmsgov/design-system-support/src/settings/index';

.ds-c-badge {
  background-color: $color-gray;
  border-radius: $border-radius-pill;
  color: $color-white;
  display: inline-block;
  font-size: $small-font-size;
  line-height: $heading-line-height;
  margin-right: $spacer-half;
  padding: $spacer-half $spacer-1;

  &:only-of-type {
    margin-right: 0;
  }
}

.ds-c-badge--info {
  background-color: $color-primary;
}

.ds-c-badge--success {
  background-color: $color-success;
}

.ds-c-badge--warn {
  background-color: $color-warn;
  color: $color-base;
}

.ds-c-badge--alert {
  background-color: $color-error;
}
