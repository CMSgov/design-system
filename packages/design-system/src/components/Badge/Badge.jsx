import PropTypes from 'prop-types';
import React from 'react';
import classNames from 'classnames';

export const Badge = props => {
  const { className, children, size, variation, ...others } = props;
  const sizeClasses = { big: 'ds-u-font-size--base' };

  const variationClass = variation && `ds-c-badge--${variation}`;
  const classes = classNames('ds-c-badge', variationClass, sizeClasses[size], className);

  return (
    <span className={classes} {...others}>
      {children}
    </span>
  );
};

Badge.propTypes = {
  /**
   * Additional classes to be added to the root badge element.
   * Useful for adding utility classes.
   */
  className: PropTypes.string,
  /**
   * Label text or HTML.
   */
  children: PropTypes.oneOfType([PropTypes.string, PropTypes.node]).isRequired,
  /**
   * Sets the font size of the Badge
   */
  size: PropTypes.oneOf(['big']),
  /**
   * A string corresponding to the badge-component variation classes
   */
  variation: PropTypes.oneOf(['info', 'success', 'warn', 'alert'])
};

export default Badge;
