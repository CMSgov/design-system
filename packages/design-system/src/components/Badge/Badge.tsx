import React from 'react';
import classNames from 'classnames';
import { t } from '../i18n';

export type BadgeSize = 'big';
export type BadgeVariation = 'info' | 'success' | 'warn' | 'alert';

export interface BadgeProps {
  /**
   * Additional classes to be added to the root badge element.
   * Useful for adding utility classes.
   */
  className?: string;
  /**
   * Label text or HTML.
   */
  children: string | React.ReactNode;
  /**
   * Sets the font size of the Badge. Only supports 'big'
   */
  size?: BadgeSize;
  /**
   * A string corresponding to the badge-component variation classes
   */
  variation?: BadgeVariation;
}

/**
 * For information about how and when to use this component,
 * [refer to its full documentation page](https://design.cms.gov/components/badge/).
 */
export const Badge: React.FC<React.ComponentPropsWithRef<'span'> & BadgeProps> = (
  props: BadgeProps
) => {
  const { className = '', children, size, variation, ...others } = props;
  const sizeClasses = { big: 'ds-c-badge--big' };

  const variationClass = variation && `ds-c-badge--${variation}`;
  const classes = classNames('ds-c-badge', variationClass, sizeClasses[size], className);

  return (
    <span className={classes} {...others}>
      {variation && (
        <span className="ds-u-visibility--screen-reader">{t(`badge.${variation}`)}: </span>
      )}
      {children}
    </span>
  );
};

export default Badge;
