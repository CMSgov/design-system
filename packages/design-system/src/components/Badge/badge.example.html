<div class="example--wrapper">
  <div>
    <span class="ds-c-badge">1</span>
    <span class="ds-c-badge">Default badge</span>
    <span class="ds-c-badge ds-u-font-size--base">Larger badge label</span>
  </div>
  <div>
    <span class="ds-c-badge ds-c-badge--info">2</span>
    <span class="ds-c-badge ds-c-badge--info">Info badge</span>
    <span class="ds-c-badge ds-c-badge--info ds-u-font-size--base">Larger info badge</span>
  </div>
  <div>
    <span class="ds-c-badge ds-c-badge--success">3</span>
    <span class="ds-c-badge ds-c-badge--success">Success badge</span>
    <span class="ds-c-badge ds-c-badge--success ds-u-font-size--base">Larger success badge</span>
  </div>
  <div>
    <span class="ds-c-badge ds-c-badge--warn">4</span>
    <span class="ds-c-badge ds-c-badge--warn">Warning badge</span>
    <span class="ds-c-badge ds-c-badge--warn ds-u-font-size--base">Larger warning badge</span>
  </div>
  <div>
    <span class="ds-c-badge ds-c-badge--alert">5</span>
    <span class="ds-c-badge ds-c-badge--alert">Alert badge</span>
    <span class="ds-c-badge ds-c-badge--alert ds-u-font-size--base">Larger alert badge</span>
  </div>
</div>
