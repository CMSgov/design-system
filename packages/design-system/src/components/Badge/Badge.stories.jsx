import React from 'react';

import Badge from './Badge';

export default {
  title: 'Components/Badge',
  component: Badge,
  argTypes: {
    size: {
      options: ['default', 'big'],
      control: { type: 'select' },
    },
    variation: {
      options: ['default', 'info', 'success', 'warn', 'alert'],
      control: { type: 'select' },
    },
  },
};

const Template = (args) => (
  <>
    <div>
      <Badge {...args}>1</Badge>
      <Badge {...args}>Badge</Badge>
      <Badge {...args} size="big">
        Badge with size &apos;big&apos;
      </Badge>
    </div>
  </>
);

export const Default = Template.bind({});
export const Alert = Template.bind({});
Alert.args = { variation: 'alert' };
export const Warn = Template.bind({});
Warn.args = { variation: 'warn' };
export const Success = Template.bind({});
Success.args = { variation: 'success' };
export const Info = Template.bind({});
Info.args = { variation: 'info' };
