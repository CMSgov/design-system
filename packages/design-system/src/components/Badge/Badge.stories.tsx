import type { Meta, StoryObj } from '@storybook/react';
import Badge from './Badge';

const meta: Meta = {
  component: Badge,
  argTypes: {
    size: {
      options: ['default', 'big'],
      control: { type: 'radio' },
    },
    variation: {
      options: ['default', 'info', 'success', 'warn', 'alert'],
      control: { type: 'radio' },
    },
    children: { control: 'text' },
  },
  parameters: {
    docs: {
      underlyingHtmlElements: ['span'],
    },
  },
};

export default meta;

type Story = StoryObj<typeof Badge>;
const BadgeTemplate: Story = {
  render: ({ ...args }) => {
    return <Badge {...args}>{args.children}</Badge>;
  },
};

export const Default = {
  ...BadgeTemplate,
  args: {
    children: '6',
  },
};

export const Informational = {
  ...BadgeTemplate,
  args: {
    variation: 'info',
    children: '29',
  },
};

export const Success = {
  ...BadgeTemplate,
  args: {
    variation: 'success',
    children: 'Best value',
  },
};

export const Warning = {
  ...BadgeTemplate,
  args: {
    variation: 'warn',
    children: 'Expiring soon',
  },
};

export const Alert = {
  ...BadgeTemplate,
  args: {
    variation: 'alert',
    children: 'Out of network',
  },
};

export const Big = {
  ...BadgeTemplate,
  args: {
    size: 'big',
    children: 'Ibuprofen',
  },
};
