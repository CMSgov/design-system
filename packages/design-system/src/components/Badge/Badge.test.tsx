import React from 'react';
import Badge from './Badge';
import { render, screen } from '@testing-library/react';

const defaultProps = {
  children: 'Foo',
};
function renderBadge(props = {}) {
  return render(<Badge {...defaultProps} {...props} />);
}

describe('Badge', () => {
  it('should render a default badge', () => {
    renderBadge();
    const badge = screen.getByText('Foo');
    expect(badge).toMatchSnapshot();
  });

  it('applies variation classes', () => {
    renderBadge({ variation: 'success' });
    const badge = screen.getByText('Foo');
    expect(badge.classList).toContain('ds-c-badge--success');
  });

  it('should render a big badge', () => {
    renderBadge({ size: 'big' });
    const badge = screen.getByText('Foo');
    expect(badge.className).toContain('ds-c-badge--big');
  });

  it('should render custom classNames', () => {
    renderBadge({ className: 'bar' });
    const badge = screen.getByText('Foo');
    expect(badge.className).toContain('bar');
  });
});
