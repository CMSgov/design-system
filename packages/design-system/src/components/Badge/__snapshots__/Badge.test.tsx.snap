// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Badge should render a default badge 1`] = `
<span
  class="ds-c-badge"
>
  Foo
</span>
`;
